// mobx-utils/lib/utils.js
export var NOOP = function () { };
export var IDENTITY = function (_) { return _; };
export function invariant(cond, message) {
    if (message === void 0) { message = "Illegal state"; }
    if (!cond)
        throw new Error("[mobx-utils] " + message);
}
var deprecatedMessages = [];
export function deprecated(msg) {
    if (deprecatedMessages.indexOf(msg) !== -1)
        return;
    deprecatedMessages.push(msg);
    console.error("[mobx-utils] Deprecated: " + msg);
}
export function addHiddenProp(object, propName, value) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: true,
        configurable: true,
        value: value
    });
}
var isGetter = function (x, name) { return (Object.getOwnPropertyDescriptor(x, name) || {}).get; };
var isFunction = function (x, name) { return typeof x[name] === "function"; };
var deepFunctions = function (x) {
    return x && x !== Object.prototype &&
        Object.getOwnPropertyNames(x)
            .filter(function (name) { return isGetter(x, name) || isFunction(x, name); })
            .concat(deepFunctions(Object.getPrototypeOf(x)) || []);
};
var distinctDeepFunctions = function (x) { return Array.from(new Set(deepFunctions(x))); };
export var getAllMethodsAndProperties = function (x) { return distinctDeepFunctions(x).filter(function (name) { return name !== "constructor" && !~name.indexOf("__"); }); };
